tests/test_distributions.py

```diff
@@ -111,3 +114,76 @@ def test_categorical(dist, CAT_ACTIONS):
+        MultiCategoricalDistribution([N_ACTIONS, N_ACTIONS]).proba_distribution(th.rand(1, sum([N_ACTIONS, N_ACTIONS]))),
+        SquashedDiagGaussianDistribution(N_ACTIONS).proba_distribution(th.rand(N_ACTIONS), th.rand(N_ACTIONS)),
+        StateDependentNoiseDistribution(N_ACTIONS).proba_distribution(
+            th.rand(N_ACTIONS), th.rand([N_ACTIONS, N_ACTIONS]), th.rand([N_ACTIONS, N_ACTIONS])
+        ),
+    ],
+)
+def test_kl_divergence(dist_type):
+    # Test 1: same distribution should have KL Div = 0
+    dist1 = dist_type
+    dist2 = dist_type
+    # PyTorch implementation of kl_divergence doesn't sum across dimensions
+    # so we need to check each one
+    assert th.all(kl_divergence(dist1, dist2).bool()) == False
+
+    # Test 2: KL Div = E(Unbiased approx KL Div)
+    if isinstance(dist_type, CategoricalDistribution):
+        dist1 = dist_type.proba_distribution(th.rand(N_ACTIONS).repeat(N_SAMPLES, 1))
+        # deepcopy needed to assign new memory to new distribution instance
+        dist2 = deepcopy(dist_type).proba_distribution(th.rand(N_ACTIONS).repeat(N_SAMPLES, 1))
+    elif isinstance(dist_type, DiagGaussianDistribution) or isinstance(dist_type, SquashedDiagGaussianDistribution):
+        mean_actions1 = th.rand(1).repeat(N_SAMPLES, 1)
+        log_std1 = th.rand(1).repeat(N_SAMPLES, 1)
+        mean_actions2 = th.rand(1).repeat(N_SAMPLES, 1)
+        log_std2 = th.rand(1).repeat(N_SAMPLES, 1)
+        dist1 = dist_type.proba_distribution(mean_actions1, log_std1)
+        dist2 = deepcopy(dist_type).proba_distribution(mean_actions2, log_std2)
+    elif isinstance(dist_type, BernoulliDistribution):
+        dist1 = dist_type.proba_distribution(th.rand(1).repeat(N_SAMPLES, 1))
+        dist2 = deepcopy(dist_type).proba_distribution(th.rand(1).repeat(N_SAMPLES, 1))
+    elif isinstance(dist_type, MultiCategoricalDistribution):
+        dist1 = dist_type.proba_distribution(th.rand(1, sum([N_ACTIONS, N_ACTIONS])).repeat(N_SAMPLES, 1))
+        dist2 = deepcopy(dist_type).proba_distribution(th.rand(1, sum([N_ACTIONS, N_ACTIONS])).repeat(N_SAMPLES, 1))
+    elif isinstance(dist_type, StateDependentNoiseDistribution):
+        dist1 = StateDependentNoiseDistribution(1)
+        dist2 = deepcopy(dist1)
+        state = th.rand(N_SAMPLES, N_FEATURES)
+        mean_actions1 = th.rand(1).repeat(N_SAMPLES, 1)
+        mean_actions2 = th.rand(1).repeat(N_SAMPLES, 1)
+        _, log_std = dist1.proba_distribution_net(N_FEATURES, log_std_init=th.log(th.tensor(0.2)))
+        dist1.sample_weights(log_std, batch_size=N_SAMPLES)
+        dist2.sample_weights(log_std, batch_size=N_SAMPLES)
+        dist1 = dist1.proba_distribution(mean_actions1, log_std, state)
+        dist2 = dist2.proba_distribution(mean_actions2, log_std, state)
+
+    full_kl_div = kl_divergence(dist1, dist2).mean(dim=0)
+    actions = dist1.get_actions()
+    approx_kl_div = (dist1.log_prob(actions) - dist2.log_prob(actions)).mean(dim=0)
```

Comment:
why not using the estimator from https://github.com/DLR-RM/stable-baselines3/pull/419 ?

Reply:
There's not much point in using the more complicated estimator since in this case we just want to assert the expected value of an unbiased estimator is equal to the original metric. For this, the variance of the metric doesn't matter so we may as well use the simpler form, using the other estimator would yield the same result just it would take longer to do the extra calculations -> no upside.